endpoints/registry.go

```diff
@@ -0,0 +1,85 @@
+package endpoints
+
+import (
+	"sync"
+	"time"
+)
+
+type Metrics interface {
+	DetectedTime() time.Time
+	InflightRequests() int64
+}
+
+type entry struct {
+	detected         time.Time
+	inflightRequests int64
+}
+
+func (e *entry) DetectedTime() time.Time {
+	return e.detected
+}
+
+func (e *entry) InflightRequests() int64 {
+	return e.inflightRequests
+}
+
+type Registry struct {
+	mu sync.Mutex
+
+	data map[string]*entry
+}
+
+type RegistryOptions struct {
+}
+
+// getOrInitEntryLocked returns pointer to endpoint registry entry
+// which contains the information about endpoint representing the
+// following key. r.mu must be held while calling this function and
+// using of the entry returned. In general, key represents the host
+// returned by normalizeSchemeHost function.
+func (r *Registry) getOrInitEntryLocked(key string) *entry {
+	e, ok := r.data[key]
+	if !ok {
+		e = &entry{}
+		r.data[key] = e
+	}
+	return e
+}
+
+func NewRegistry(o RegistryOptions) *Registry {
+	return &Registry{data: map[string]*entry{}}
+}
+
+func (r *Registry) GetMetrics(key string) Metrics {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	e := r.getOrInitEntryLocked(key)
+	copy := &entry{}
+	*copy = *e
```

Comment:
No need to copy as there is no way to mutate value via Metrics interface

Reply:
But there is a way to mutate: if user of `endpoints.Registry` will use `endpoint.Registry.SetDetectedTime` with the same key, it will get the same pointer and will overwrite it, for the user called `endpoint.Registry.GetMetrics` as well.